File: .leetcode/40.组合总和-ii.cpp
/*
 * @lc app=leetcode.cn id=40 lang=cpp
 *
 * [40] 组合总和 II
 */
#include<bits/stdc++.h>
using namespace std;

// @lc code=start
class Solution {
public:
    vector<vector<int>>ans;
    vector<int>temp;
    vector<pair<int,int>>freq;
    void traversal(int begin,int target){
        if(target==0){
            ans.push_back(temp);
            return;
        }
        if(begin == freq.size()||target < freq[begin].first)return;
        traversal(begin + 1,target);
        int most = min(target/freq[begin].first,freq[begin].second);
        for(int i = 1;i <= most;i++){
            temp.push_back(freq[begin].first);
            traversal(begin + 1,target - i * freq[begin].first);
        }
        for(int i = 0;i < most;i++)temp.pop_back();
    }
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        for(auto i:candidates){
            if(freq.empty()||freq.back().first!=i)
                freq.emplace_back(i,1);
            else
                freq.back().second++;
        }
        traversal(0,target);
        return ans;
    }
};
// @lc code=end

File: .leetcode/40.组合总和-ii.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

/*
 * @lc app=leetcode.cn id=40 lang=java
 *
 * [40] 组合总和 II
 */

// @lc code=start
class Solution {
    List<List<Integer>> ans = new LinkedList<>();
    List<Integer>temp = new LinkedList<>();
    int sum;
    public void dfs(int start,int[][]a,int target){
        if(sum == target){
            ans.add(new LinkedList<>(temp));
            return;
        }
        if(sum > target || start >= a.length)return;
        dfs(start + 1, a, target);
        int n = (target - sum) / a[start][0];
        int t = Math.min(n, a[start][1]);
        sum += t * a[start][0];
        for(int i = 0;i < t;i++){
            temp.add(a[start][0]);
        }
        for(int i = 0;i < t;i++){
            dfs(start + 1, a, target);
            temp.remove(temp.size() - 1);
            sum -= a[start][0];
        }
    }
    public List<List<Integer>> combinationSum2(int[] candidates, int target) {
        Map<Integer,Integer>mp = new HashMap();
        for(int i : candidates){
            mp.put(i, mp.getOrDefault(i, 0) + 1);
        }
        int[][] a = new int[mp.size()][2];
        int index = 0;
        for(Map.Entry<Integer,Integer> i : mp.entrySet()){
            a[index][0] = i.getKey();
            a[index++][1] = i.getValue();
        }
        dfs(0, a, target);
        return ans;
    }
}
// @lc code=end

